From 134f6ba580ebc7b84169671ceca794e47639b5e2 Mon Sep 17 00:00:00 2001
From: Christian Hohnstaedt <Christian.Hohnstaedt@wago.com>
Date: Tue, 17 Sep 2019 08:05:42 +0200
Subject: [PATCH] rmd: add debugfs entries for the buffer control words

Add debugfs entries for "tripple_buf_ctrl_[r|t]x",
displaying the current state of the buffer control word.

Signed-off-by: Christian Hohnstaedt <Christian.Hohnstaedt@wago.com>
---
 drivers/char/rmd.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/char/rmd.c b/drivers/char/rmd.c
index f3c998e..4017937 100644
--- a/drivers/char/rmd.c
+++ b/drivers/char/rmd.c
@@ -17,6 +17,8 @@
 #include <linux/uaccess.h>
 #include <linux/rmd.h>
 
+#include <linux/debugfs.h>
+
 #define RMD_NAME "rmd"
 #define RMD_MAX_DEVICES 42
 #define RMD_TRIPPLE_BUF_SIZE 3
@@ -94,6 +96,7 @@ struct rmd {
 	 * next cycle
 	 */
 	ktime_t			last_call_time;
+	struct dentry		*debug;
 };
 
 struct rmd_info {
@@ -495,6 +498,8 @@ static void rmd_cleanup(struct rmd_info *info)
 	if (!info->is_configured)
 		return;
 
+	debugfs_remove_recursive(rmd->debug);
+
 	info->is_configured = false;
 	info->is_tx_enabled = false;
 
@@ -570,9 +575,16 @@ static int rmd_setup(struct rmd_info *info, void __user *argp)
 		ret = rmd_dma_setup(info);
 	}
 
-	if (!ret)
-		info->is_configured = true;
+	if (ret)
+		return ret;
+
+	info->is_configured = true;
 
+	rmd->debug = debugfs_create_dir(dev_name(rmd->dev), NULL);
+	debugfs_create_x32("tripple_buf_ctrl_rx", 0600, rmd->debug,
+				info->tripple_buf_ctrl);
+	debugfs_create_x32("tripple_buf_ctrl_tx", 0600, rmd->debug,
+				info->tripple_buf_ctrl + sizeof(u32));
 	return ret;
 }
 
-- 
2.7.4

